handlers/whatsapp/whatsapp.go

```diff
@@ -0,0 +1,282 @@
+package telegram
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"net/url"
+	"strconv"
+	"time"
+
+	"github.com/buger/jsonparser"
+	"github.com/nyaruka/courier"
+	"github.com/nyaruka/courier/handlers"
+	"github.com/nyaruka/courier/utils"
+	"github.com/nyaruka/gocommon/urns"
+	"github.com/pkg/errors"
+)
+
+func init() {
+	courier.RegisterHandler(NewHandler())
+}
+
+type handler struct {
+	handlers.BaseHandler
+}
+
+// NewHandler returns a new WhatsApp ready to be registered
+func NewHandler() courier.ChannelHandler {
+	return &handler{handlers.NewBaseHandler(courier.ChannelType("WA"), "WhatsApp")}
+}
+
+// Initialize is called by the engine once everything is loaded
+func (h *handler) Initialize(s courier.Server) error {
+	h.SetServer(s)
+	err := s.AddReceiveMsgRoute(h, http.MethodPost, "receive", h.ReceiveMessage)
+	if err != nil {
+		return err
+	}
+	return s.AddUpdateStatusRoute(h, http.MethodPost, "status", h.UpdateStatus)
+}
+
+// {
+//	 "meta": null,
+//	 "payload": {
+//	   "from": "16315555555",
+//	   "message_id": "345b5e14775782",
+//	   "timestamp": "1476225801",
+//	   "message": {
+//		"address": "1 Hacker Way, Menlo Park, CA 94025",
+//		"latitude": 37.483253479003906,
+//		"longitude": -122.14960479736328,
+// 		"has_media": true,
+// 		"text": "This is the media caption.",
+//		"type": "image"
+// 	   }
+//	 },
+//	 "error": false
+// }
+type waReceive struct {
+	Payload struct {
+		From      string `json:"from" validate:"required"`
+		MessageID string `json:"message_id" validate:"required"`
+		Timestamp string `json:"timestamp" validate:"required"`
+		Message   struct {
+			HasMedia  bool    `json:"has_media"`
+			Address   string  `json:"address"`
+			Latitude  float64 `json:"latitude"`
+			Longitude float64 `json:"longitude"`
```

Comment:
It seems there are missing fields here for the media case but I saw the TODOs

Reply:
Media is super weird in that it will actually come in a totally different request.. but no, these are all the fields present / possible. But ya, not implemented fully yet.